### ISSUE TITLE
FlexAttention: create_block_mask is considerably slow when using flattened 1D sequences (document masking / jagged tensors)

### ISSUE THREAD

**Issue by mikkelfo** [pytorch/pytorch#151043]
### 🐛 Describe the bug

FlexAttention seems considerably slow for 1D-vector approaches where the batch_size and seq_len is collapsed. This scenario is pretty common for any document_masking or jagged tensors setups and is even part of the [FlexAttention blog examples](https://pytorch.org/blog/flexattention/#document-maskingjagged-sequences). Especially prevalent when chaining together multiple `mask_mod` (to force re-compilation every time), then `create_block_mask` runs significantly slower than I would expect. On various systems, the below code takes between 0.15-0.30s per call, which seems quite slow.

Do I have an unreasonable expectation of the performance of a 1D `(batch_size*seq_len)` FlexAttention implementation, that has to re-compile every batch (document masking/jagged tensor setups)? 
```
import time
import torch
from torch.nn.attention.flex_attention import create_block_mask, and_masks

def wrapper(randn):
    def overzero(b, h, q_idx, kv_idx):
        return randn[q_idx] >= 0
    return overzero
def causal(b, h, q_idx, kv_idx):
    return q_idx >= kv_idx

batch_size, seq_len = 512, 512

# Warmup
randn = torch.randn(batch_size*seq_len, device=torch.device("cuda"))
block_mask = create_block_mask(and_masks(wrapper(randn), causal), None, None, batch_size*seq_len, batch_size*seq_len, _compile=True)

total = 0
n = 50
for _ in range(n):
    randn = torch.randn(batch_size*seq_len, device=torch.device("cuda"))
    start = time.perf_counter()
    block_mask = create_block_mask(and_masks(wrapper(randn), causal), None, None, batch_size*seq_len, batch_size*seq_len, _compile=True)
    end = time.perf_counter()

    total += end - start
print(total/n)
```

### Versions

Seen on both 2.5.1 and 2.6

**Comment by drisspg** [comment]
Yeah this is going to be super slow, you have to compile the create_block_mask_function. e.g. `cbm_compiled ` torch.compile(create_block_mask)`

**Comment by mikkelfo** [comment]
@drisspg Unless I'm mistaken, this happens internally inside using the `_compile=True` flag in versions 2.5 and 2.6. (see https://github.com/pytorch/pytorch/blob/f3cf3ec591528e1fd4b56dba6da5a0d803c61dc9/torch/nn/attention/flex_attention.py#L880-L887), so I don't see that this should be the issue. 

Running the above code without `_compile=True` also yields absurdly high memory usage and often ends in OOM errors (64GiB for the above example)

**Comment by drisspg** [comment]
Sorry was reading from my phone and missed it;
```Py
import time
import torch
from torch.nn.attention.flex_attention import create_block_mask, and_masks
import transformer_nuggets
from transformer_nuggets.utils.benchmark import profiler, benchmark_cuda_function_in_microseconds

# Random wrapper function to mimic document_masking setup and to force recompilation  
def wrapper(randn): 
    def overzero(b, h, q_idx, kv_idx):
        return randn[q_idx] >= 0
    return overzero
def causal(b, h, q_idx, kv_idx):
    return q_idx >= kv_idx

batch_size, seq_len = 512, 512

# Warmup
randn = torch.randn(batch_size*seq_len, device=torch.device("cuda"))
cbm_compile = torch.compile(create_block_mask, fullgraph=True, mode="max-autotune-no-cudagraphs")
final_mask = and_masks(wrapper(randn), causal)
total_seq_len = batch_size*seq_len

block_mask = cbm_compile(final_mask, None, None, total_seq_len, total_seq_len)

total = 0
n = 50
for _ in range(n):
    randn = torch.randn(total_seq_len, device=torch.device("cuda"))
    start = time.perf_counter()
    block_mask = cbm_compile(final_mask, None, None, total_seq_len, total_seq_len)
    end = time.perf_counter()

    total += end - start
print(total/n)

cuda_time = benchmark_cuda_function_in_microseconds(cbm_compile, final_mask, None, None, total_seq_len, total_seq_len)
print(f"CUDA time: {cuda_time}us")

with profiler("data/mikk"):
    block_mask = cbm_compile(final_mask, None, None, total_seq_len, total_seq_len)

# Each buffer has 2048*2048 = 4,194,304 elements
# Each element is 4 bytes (int32)
# Two buffers write
# 1 buffer read
write_size = (((total_seq_len/128)**2) * 4) * 3
# cuda time in seconds
cuda_time = cuda_time * 1e-6 # seconds
print(f"Write size: {write_size} bytes")
write_throughput = write_size * 1e-9 / cuda_time # bytes per second
print(f"Write throughput: {write_throughput} GB/s")

```

Does in fact have 1 slow kernel that takes around 31 ms on my machine
```
❯ python flex/mikk_repro.py
0.013150181267410516
CUDA time: 32861.53699999999us
transformer_nuggets.utils.benchmark - INFO - 💾 Trace file 📄 saved to: data/mikk.json
Write size: 50331648.0 bytes
Write throughput: 1.5316279332886962 GB/s
```

This is the large reduction kernel
```Py
@triton.jit
def triton_red_fused__to_copy_bitwise_and_eq_gt_lt_sum_0(in_ptr0, out_ptr1, out_ptr2, xnumel, r0_numel, XBLOCK : tl.constexpr, R0_BLOCK : tl.constexpr):
    xnumel = 4194304
    r0_numel = 16384
    rnumel = r0_numel
    RBLOCK: tl.constexpr = R0_BLOCK
    xoffset = tl.program_id(0).to(tl.int64) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None].to(tl.int64)
    xmask = tl.full([XBLOCK, R0_BLOCK], True, tl.int1)
    r0_base = tl.arange(0, R0_BLOCK)[None, :].to(tl.int64)
    rbase = r0_base
    x1 = xindex // 2048
    x0 = (xindex % 2048)
    _tmp11 = tl.full([XBLOCK, R0_BLOCK], 0, tl.int64)
    x4 = xindex
    for r0_offset in range(0, r0_numel, R0_BLOCK):
        r0_index = r0_offset + r0_base
        r0_mask = r0_index < r0_numel
        roffset = r0_offset
        rindex = r0_index
        r0_3 = r0_index // 128
        r0_2 = (r0_index % 128)
        tmp0 = tl.load(in_ptr0 + (r0_3 + 128*x1), r0_mask, eviction_policy='evict_last', other=0.0)
        tmp1 = 0.0
        tmp2 = tmp0 >= tmp1
        tmp3 = tl.full([1, 1], True, tl.int1)
        tmp4 = tmp3 & tmp2
        tmp5 = r0_3 + 128*x1
        tmp6 = r0_2 + 128*x0
        tmp7 = tmp5 >= tmp6
        tmp8 = tmp4 & tmp7
        tmp9 = tmp8.to(tl.int64)
        tmp10 = tl.broadcast_to(tmp9, [XBLOCK, R0_BLOCK])
        tmp12 = _tmp11 + tmp10
        _tmp11 = tl.where(r0_mask, tmp12, _tmp11)
    tmp11 = tl.sum(_tmp11, 1)[:, None]
    tmp13 = tl.full([1, 1], 0, tl.int64)
    tmp14 = tmp11 > tmp13
    tmp15 = tl.full([1, 1], 16384, tl.int64)
    tmp16 = tmp11 < tmp15
    tmp17 = tmp14 & tmp16
    tmp18 = tmp17.to(tl.int8)
    tmp19 = tmp18.to(tl.int32)
    tmp20 = tmp11 == tmp15
    tmp21 = tmp20.to(tl.int8)
    tmp22 = tmp21.to(tl.int32)
    tl.store(out_ptr1 + (x4), tmp19, None)
    tl.store(out_ptr2 + (x4), tmp22, None)
''', device_str='cuda')
```

<img width="1148" alt="Image" src="https://github.com/user-attachments/assets/8a9611c2-4bc3-4ec8-b6ea-3f61a1ae2f3c" />

will continue to investigate the perf of this and if there are ways to improve

**Comment by mikkelfo** [comment]
@drisspg Looks like yours takes 0.013 seconds which is more reasonable, despite the slow kernel you mentioned. I've tried on multiple systems and sometimes I'm "lucky" to get down to 0.05s (even on a T4), but on my production setup (A100) it seems to sit at 0.3s (~0.15s with the above code) consistently as I'm combining 3 `mask_mod` using `and_masks`

Any ideas about what causes the large deviations in run time, especially when combining multiple `mask_mod`?

**Comment by drisspg** [comment]
Well I am using an H100 so that might explain a large execution difference between mine and yours

**Comment by Chillee** [comment]
@mikkelfo Could you perhaps try doing `torch.compile(mode="max-autotune-no-cudagraphs")`? I wonder if it's a block size tuning issue.

**Comment by mikkelfo** [comment]
@drisspg Can you try adding the below code to your profiling, as this dramatically slows down the `create_block_mask` by 4x. I wonder if it makes any potential issues more apparent
```
def doc_mask(mask):
    def doc(b, h, q_idx, kv_idx):
        return mask[q_idx] == mask[kv_idx]
    return doc
mask = torch.tensor([[i]*seq_len for i in range(batch_size)], device=torch.device("cuda")).flatten()
final_mask = and_masks(wrapper(randn), causal, doc_mask(mask)
```

@Chillee using your suggestion, the runtime goes from 0.12s -> 0.077s on my system. Adding the `doc_mask` function from above, a standard `torch.compile(create_block_mask)` takes 0.44s, and using `mode="max-autotune-no-cudagraphs"` reduces this to 0.37s.

**Comment by mikkelfo** [comment]
Another thing I found, which I find odd, is that the combination of your `mask_mods` appears to matter, and combining them into a single one results in some non-negligible speedup:
The below results use 2.6, `torch.compile(mode="max-autotune-no-cudagraphs")` and run in colab on a T4. 

`and_masks(wrapper(randn), causal, doc_mask(mask) -> 0.37s`

If I combine the `causal` with e.g. `doc_mask` like so
```
def doc_mask(mask):
    def doc(b, h, q_idx, kv_idx):
        return (mask[q_idx] == mask[kv_idx]) & (q_idx >= kv_idx)
    return doc
```
`and_masks(wrapper(randn), doc_mask(mask) -> 0.31s`

And finally, if I manually combine all three `mask_mods` without using `and_masks`
```
def combined(mask, randn):
    def comb(b, h, q_idx, kv_idx):
        return (mask[q_idx] == mask[kv_idx]) & (randn[q_idx] >= 0) & (q_idx >= kv_idx)
    return comb
```
`combined(mask, randn) -> 0.30s`

Is this just a side-effect of how `and_masks` is implemented, where result is updated for each mask, which I guess is more robust than simply doing `mask_mod1 & mask_mod2 & mask_mod3`?
https://github.com/pytorch/pytorch/blob/f3cf3ec591528e1fd4b56dba6da5a0d803c61dc9/torch/nn/attention/flex_attention.py#L726-L730

**Comment by aorenste** [comment]
Performance issue with no concrete pass/fail criterion — the snippet runs successfully and only measures wall-clock time, which varies substantially by GPU (0.013s on H100 vs 0.3s on A100 per the thread)

Specifically:
- No expected error or assertion — this is a perf characterization, not a crash or correctness bug. To be auto-reproducible we'd need a concrete threshold (e.g. 'create_block_mask should take < Xs on hardware Y') that can raise AssertionError when exceeded.
- Reported timings vary dramatically across GPUs in the thread itself (H100 ~0.013s, T4/A100 ~0.15–0.4s), so any threshold needs to be pinned to a specific platform; the issue doesn't state which platform is canonical for the regression.
- The thread is partly exploratory (comparing and_masks vs hand-combined mask_mods, varying mode= settings) without converging on a single 'this is the bug' configuration — needs a single minimal repro plus the specific time/throughput regression being claimed.

Please add a self-contained script (or instructions to obtain any missing dependencies) so the team can investigate.

_(Analysis done by Claude.)_

**Comment by drisspg** [comment]
I am going to close this issue since this was a few verisons ago and we have changed triton a number of times, feel free to reopen with more context if still relevant
